src/lib/operators/join_sort_merge/column_materializer.hpp

```diff
@@ -67,24 +93,51 @@ class ColumnMaterializer {
    **/
   std::shared_ptr<AbstractTask> _create_chunk_materialization_job(std::unique_ptr<MaterializedSegmentList<T>>& output,
                                                                   std::unique_ptr<PosList>& null_rows_output,
-                                                                  ChunkID chunk_id, std::shared_ptr<const Table> input,
-                                                                  ColumnID column_id) {
-    return std::make_shared<JobTask>([this, &output, &null_rows_output, input, column_id, chunk_id] {
+                                                                  const ChunkID chunk_id,
+                                                                  std::shared_ptr<const Table> input,
+                                                                  const ColumnID column_id,
+                                                                  SampleRequest<T>& sample_request) {
+    return std::make_shared<JobTask>([this, &output, &null_rows_output, input, column_id, chunk_id, &sample_request] {
       auto segment = input->get_chunk(chunk_id)->get_segment(column_id);
 
       if (const auto dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<T>>(segment)) {
-        (*output)[chunk_id] = _materialize_dictionary_segment(*dictionary_segment, chunk_id, null_rows_output);
+        (*output)[chunk_id] =
+            _materialize_dictionary_segment(*dictionary_segment, chunk_id, null_rows_output, sample_request);
       } else {
-        (*output)[chunk_id] = _materialize_generic_segment(*segment, chunk_id, null_rows_output);
+        (*output)[chunk_id] = _materialize_generic_segment(*segment, chunk_id, null_rows_output, sample_request);
       }
     });
   }
 
+  /**
+   * Samples values from a materialized segment.
+   * We collect locally write once to the global sample collection to limit non-local writes.
+   */
+  void _gather_samples_from_segment(MaterializedSegment<T>& segment_output, SampleRequest<T>& sample_request) const {
```

Comment:
i don't like the signature of this. On first sight, I thought that the first parameter was the input.